Intermittent test failure: test_check_with_rejected_launch_rolls_back_and_pushes

Seen locally


```
TestNimbusCheckKintoPushQueueByCollection.test_check_with_rejected_launch_rolls_back_and_pushes _
[gw3] linux -- Python 3.9.6 /usr/local/bin/python3

self = <experimenter.kinto.tests.test_tasks.TestNimbusCheckKintoPushQueueByCollection testMethod=test_check_with_rejected_launch_rolls_back_and_pushes>

    def test_check_with_rejected_launch_rolls_back_and_pushes(self):
        rejected_experiment = NimbusExperimentFactory.create_with_lifecycle(
            NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_WAITING,
            application=NimbusExperiment.Application.DESKTOP,
        )
        launching_experiment = NimbusExperimentFactory.create_with_lifecycle(
            NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE,
            application=NimbusExperiment.Application.DESKTOP,
        )
    
        self.setup_kinto_get_main_records([])
        self.setup_kinto_rejected_review()
    
        tasks.nimbus_check_kinto_push_queue_by_collection(
            settings.KINTO_COLLECTION_NIMBUS_DESKTOP
        )
    
        self.mock_push_task.assert_called_with(
            settings.KINTO_COLLECTION_NIMBUS_DESKTOP, launching_experiment.id
        )
        self.mock_kinto_client.patch_collection.assert_called_with(
            id=settings.KINTO_COLLECTION_NIMBUS_DESKTOP,
            bucket=settings.KINTO_BUCKET_WORKSPACE,
            data={"status": KINTO_ROLLBACK_STATUS},
        )
    
        rejected_experiment = NimbusExperiment.objects.get(id=rejected_experiment.id)
        self.assertEqual(rejected_experiment.status, NimbusExperiment.Status.DRAFT)
        self.assertEqual(
            rejected_experiment.publish_status, NimbusExperiment.PublishStatus.IDLE
        )
        self.assertIsNone(rejected_experiment.status_next)
    
>       self.assertTrue(
            rejected_experiment.changes.filter(
                changed_by__email=settings.KINTO_DEFAULT_CHANGELOG_USER,
                old_status=NimbusExperiment.Status.DRAFT,
                old_publish_status=NimbusExperiment.PublishStatus.WAITING,
                new_status=NimbusExperiment.Status.DRAFT,
                new_publish_status=NimbusExperiment.PublishStatus.IDLE,
            ).exists()
        )
E       AssertionError: False is not true

experimenter/kinto/tests/test_tasks.py:327: AssertionError
```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-1695)

--- a/app/experimenter/base/management/commands/load_dummy_experiments.py
+++ b/app/experimenter/base/management/commands/load_dummy_experiments.py
@@ -22,7 +22,5 @@ class Command(BaseCommand):
             logger.info("Created {}: {}".format(experiment, status))
 
         for lifecycle in NimbusExperimentFactory.LocalLifecycles:
-            experiment = NimbusExperimentFactory.create_with_lifecycle(
-                lifecycle, with_random_timespan=True
-            )
+            experiment = NimbusExperimentFactory.create_with_lifecycle(lifecycle)
             logger.info("Created {}: {}".format(experiment, lifecycle))

--- a/app/experimenter/experiments/tests/api/v5/test_queries.py
+++ b/app/experimenter/experiments/tests/api/v5/test_queries.py
@@ -473,7 +473,6 @@ class TestNimbusExperimentBySlugQuery(GraphQLTestCase):
         user_email = "user@example.com"
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.CREATED,
-            with_random_timespan=True,
         )
 
         response = self.query(
@@ -500,7 +499,6 @@ class TestNimbusExperimentBySlugQuery(GraphQLTestCase):
         user_email = "user@example.com"
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_REJECT,
-            with_random_timespan=True,
         )
         response = self.query(
             """
@@ -528,7 +526,6 @@ class TestNimbusExperimentBySlugQuery(GraphQLTestCase):
         user_email = "user@example.com"
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.CREATED,
-            with_random_timespan=True,
         )
 
         response = self.query(
@@ -555,7 +552,6 @@ class TestNimbusExperimentBySlugQuery(GraphQLTestCase):
         user_email = "user@example.com"
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_REVIEW_REQUESTED,
-            with_random_timespan=True,
         )
         response = self.query(
             """
@@ -583,7 +579,6 @@ class TestNimbusExperimentBySlugQuery(GraphQLTestCase):
         user_email = "user@example.com"
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_WAITING,
-            with_random_timespan=True,
         )
         response = self.query(
             """
@@ -609,7 +604,6 @@ class TestNimbusExperimentBySlugQuery(GraphQLTestCase):
         user_email = "user@example.com"
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_TIMEOUT,
-            with_random_timespan=True,
         )
         response = self.query(
             """

--- a/app/experimenter/experiments/tests/factories/nimbus.py
+++ b/app/experimenter/experiments/tests/factories/nimbus.py
@@ -296,9 +296,9 @@ class NimbusExperimentFactory(factory.django.DjangoModelFactory):
             self.countries.add(*extracted)
 
     @classmethod
-    def create_with_lifecycle(cls, lifecycle, with_random_timespan=False, **kwargs):
+    def create_with_lifecycle(cls, lifecycle, with_latest_change_now=False, **kwargs):
         experiment = cls.create(**kwargs)
-        now = timezone.now() - datetime.timedelta(days=random.randint(100, 200))
+        current_datetime = timezone.datetime(2021, 1, 1)
 
         for state in lifecycle.value:
             experiment.apply_lifecycle_state(state)
@@ -321,10 +321,14 @@ class NimbusExperimentFactory(factory.django.DjangoModelFactory):
                 f"set lifecycle {lifecycle} state {state}",
             )
 
-            if with_random_timespan:
-                change.changed_on = now
-                change.save()
-                now += datetime.timedelta(days=random.randint(5, 20))
+            change.changed_on = current_datetime
+            change.save()
+            current_datetime += datetime.timedelta(days=1)
+
+        if with_latest_change_now:
+            latest_change = experiment.changes.latest_change()
+            latest_change.changed_on = timezone.now()
+            latest_change.save()
 
         return NimbusExperiment.objects.get(id=experiment.id)
 

--- a/app/experimenter/experiments/tests/test_models/test_models_nimbus.py
+++ b/app/experimenter/experiments/tests/test_models/test_models_nimbus.py
@@ -420,6 +420,7 @@ class TestNimbusExperiment(TestCase):
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_duration=10,
+            with_latest_change_now=True,
         )
         self.assertEqual(
             experiment.proposed_end_date,
@@ -430,6 +431,7 @@ class TestNimbusExperiment(TestCase):
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_duration=10,
+            with_latest_change_now=True,
         )
         self.assertFalse(experiment.should_end)
 
@@ -450,6 +452,7 @@ class TestNimbusExperiment(TestCase):
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=10,
+            with_latest_change_now=True,
         )
         self.assertFalse(experiment.should_end_enrollment)
 
@@ -473,6 +476,10 @@ class TestNimbusExperiment(TestCase):
             old_status=NimbusExperiment.Status.DRAFT,
             new_status=NimbusExperiment.Status.LIVE,
         ).update(changed_on=datetime.datetime.now() - datetime.timedelta(days=3))
+
+        experiment.changes.filter(experiment_data__is_paused=True).update(
+            changed_on=datetime.datetime.now()
+        )
 
         self.assertEqual(
             experiment.computed_enrollment_days,
@@ -499,6 +506,11 @@ class TestNimbusExperiment(TestCase):
             old_status=NimbusExperiment.Status.DRAFT,
             new_status=NimbusExperiment.Status.LIVE,
         ).update(changed_on=datetime.datetime.now() - datetime.timedelta(days=7))
+
+        experiment.changes.filter(
+            old_status=NimbusExperiment.Status.LIVE,
+            new_status=NimbusExperiment.Status.COMPLETE,
+        ).update(changed_on=datetime.datetime.now())
 
         self.assertEqual(
             experiment.computed_duration_days,
@@ -674,6 +686,7 @@ class TestNimbusExperiment(TestCase):
         experiment = NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=10,
+            with_latest_change_now=True,
         )
         self.assertEqual(
             experiment.proposed_enrollment_end_date,
@@ -906,7 +919,9 @@ class TestNimbusExperiment(TestCase):
     )
     def test_should_timeout(self, expected, timeout, lifecycle):
         with override_settings(KINTO_REVIEW_TIMEOUT=timeout):
-            experiment = NimbusExperimentFactory.create_with_lifecycle(lifecycle)
+            experiment = NimbusExperimentFactory.create_with_lifecycle(
+                lifecycle, with_latest_change_now=True
+            )
             self.assertEqual(experiment.should_timeout, expected)
 
     def test_clone_created_experiment(self):

--- a/app/experimenter/kinto/tests/test_tasks.py
+++ b/app/experimenter/kinto/tests/test_tasks.py
@@ -100,6 +100,7 @@ class TestNimbusCheckKintoPushQueueByCollection(MockKintoClientMixin, TestCase):
         NimbusExperimentFactory.create_with_lifecycle(
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_WAITING,
             application=NimbusExperiment.Application.DESKTOP,
+            with_latest_change_now=True,
         )
 
         self.setup_kinto_get_main_records([])
@@ -781,7 +782,7 @@ class TestNimbusSynchronizePreviewExperimentsInKinto(MockKintoClientMixin, TestC
             tasks.nimbus_synchronize_preview_experiments_in_kinto()
 
 
-class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
+class TestNimbusSendEmails(MockKintoClientMixin, TestCase):
     def test_enrollment_ending_email_not_sent_for_experiments_before_enrollment_end_date(
         self,
     ):
@@ -789,6 +790,7 @@ class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=10,
             proposed_duration=20,
+            with_latest_change_now=True,
         )
 
         tasks.nimbus_send_emails()
@@ -801,6 +803,7 @@ class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=0,
             proposed_duration=20,
+            with_latest_change_now=True,
         )
 
         NimbusEmail.objects.create(
@@ -819,6 +822,7 @@ class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=0,
             proposed_duration=20,
+            with_latest_change_now=True,
         )
 
         tasks.nimbus_send_emails()
@@ -838,6 +842,7 @@ class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=10,
             proposed_duration=20,
+            with_latest_change_now=True,
         )
 
         tasks.nimbus_send_emails()
@@ -850,6 +855,7 @@ class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=10,
             proposed_duration=0,
+            with_latest_change_now=True,
         )
 
         NimbusEmail.objects.create(
@@ -868,6 +874,7 @@ class TestNimbusSendEndEnrollmentEmail(MockKintoClientMixin, TestCase):
             NimbusExperimentFactory.Lifecycles.LAUNCH_APPROVE_APPROVE,
             proposed_enrollment=10,
             proposed_duration=0,
+            with_latest_change_now=True,
         )
 
         tasks.nimbus_send_emails()

